File: Training_Content/labs/lab05-multi-agent/python/llm_client.py
"""LLM Client abstraction."""
import os
from typing import List, Dict, Optional
from abc import ABC, abstractmethod


class LLMClient(ABC):
    """Abstract base class for LLM clients."""

    @abstractmethod
    def chat(self, messages: List[Dict[str, str]]) -> str:
        """Send messages and get a response."""
        pass


class AnthropicClient(LLMClient):
    """Anthropic Claude client."""

    def __init__(self, model: str = "claude-3-5-sonnet-20241022"):
        from anthropic import Anthropic
        self.client = Anthropic()
        self.model = model

    def chat(self, messages: List[Dict[str, str]]) -> str:
        # Extract system message
        system = None
        filtered = []
        for msg in messages:
            if msg["role"] == "system":
                system = msg["content"]
            else:
                filtered.append(msg)

        response = self.client.messages.create(
            model=self.model,
            max_tokens=4096,
            system=system,
            messages=filtered
        )

        return response.content[0].text


class OpenAIClient(LLMClient):
    """OpenAI GPT client."""

    def __init__(self, model: str = "gpt-4o"):
        from openai import OpenAI
        self.client = OpenAI()
        self.model = model

    def chat(self, messages: List[Dict[str, str]]) -> str:
        response = self.client.chat.completions.create(
            model=self.model,
            messages=messages
        )
        return response.choices[0].message.content


def get_llm_client(provider: str = "anthropic") -> LLMClient:
    """Get an LLM client by provider name."""
    if provider == "anthropic":
        return AnthropicClient()
    elif provider == "openai":
        return OpenAIClient()
    else:
        raise ValueError(f"Unknown provider: {provider}")

File: Training_Content/labs/lab04-rag-system/typescript/src/llm-client.ts
/**
 * LLM Client Abstraction
 */

export interface Message {
  role: 'system' | 'user' | 'assistant';
  content: string;
}

export abstract class LLMClient {
  abstract chat(messages: Message[]): Promise<string>;
}

export class AnthropicClient extends LLMClient {
  private client: InstanceType<typeof import('@anthropic-ai/sdk').default> | null = null;
  private model: string;

  constructor(model: string = 'claude-3-5-sonnet-20241022') {
    super();
    this.model = model;
  }

  private async ensureClient(): Promise<void> {
    if (!this.client) {
      const Anthropic = (await import('@anthropic-ai/sdk')).default;
      this.client = new Anthropic();
    }
  }

  async chat(messages: Message[]): Promise<string> {
    await this.ensureClient();

    let system: string | undefined;
    const filtered: Array<{ role: 'user' | 'assistant'; content: string }> = [];

    for (const m of messages) {
      if (m.role === 'system') {
        system = m.content;
      } else {
        filtered.push({ role: m.role, content: m.content });
      }
    }

    const response = await this.client!.messages.create({
      model: this.model,
      max_tokens: 4096,
      system,
      messages: filtered,
    });

    const textBlock = response.content.find((block) => block.type === 'text');
    return textBlock?.type === 'text' ? textBlock.text : '';
  }
}

export class OpenAIClient extends LLMClient {
  private client: InstanceType<typeof import('openai').default> | null = null;
  private model: string;

  constructor(model: string = 'gpt-4o') {
    super();
    this.model = model;
  }

  private async ensureClient(): Promise<void> {
    if (!this.client) {
      const OpenAI = (await import('openai')).default;
      this.client = new OpenAI();
    }
  }

  async chat(messages: Message[]): Promise<string> {
    await this.ensureClient();

    const response = await this.client!.chat.completions.create({
      model: this.model,
      messages,
    });

    return response.choices[0].message.content || '';
  }
}

export type LLMProvider = 'anthropic' | 'openai';

export function getLLMClient(provider: LLMProvider = 'anthropic'): LLMClient {
  switch (provider) {
    case 'anthropic':
      return new AnthropicClient();
    case 'openai':
      return new OpenAIClient();
    default:
      throw new Error(`Unknown provider: ${provider}`);
  }
}
